### PR TITLE
Adds support for CAA record types

### PR DESCRIPTION
**Pull request by danvaida**
Upcoming release: `v0.2.0`

### CODE REVIEW

File: modules/record_collection_data/outputs.tf [hunk 1]
```diff
@@ -0,0 +1,9 @@
+output "values" {
+  description = "Cloudflare Zone DNS Records"
+  value = {
+    "${cloudflare_record.this_data.data[0].flags} ${cloudflare_record.this_data.data[0].tag} ${cloudflare_record.this_data.data[0].value}" = {
```
Comment (snovikov) [inline modules/record_collection_data/outputs.tf:4]: @danvaida could we probably use underscores(`_`) here as separators? Otherwise it looks kind of wierd in the output:

```
      + CAA_somedomain.example.com        = {
          + 0 issue someca.com = {
              + created = (known after apply)
              + id      = (known after apply)
            }
        }
```

File: main.tf [hunk 2]
```diff
@@ -20,3 +23,15 @@ module "records" {
...
 }
+
+module "records_data" {
+  source   = "./modules/record_collection_data"
+  for_each = local.collections_data
+
+  zone_id = data.cloudflare_zone.this.id
+  name    = each.value[0].name
+  type    = each.value[0].type
+  data    = each.value[0].data
```
Comment (snovikov) [inline main.tf:34]: You have to iterate here. In case you have multiple records with different data blocks.

File: modules/record_collection_data/main.tf [hunk 3]
```diff
@@ -0,0 +1,13 @@
+resource "cloudflare_record" "this_data" {
```
Comment (snovikov) [inline modules/record_collection_data/main.tf:1]: You have to iterate here in the same way as in `record_collection` submodule.

File: modules/record_collection_data/variables.tf [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+  description = "Cloudflare Hosted Zone Record Name"
+  type        = string
+}
+
+variable "type" {
+  description = "Cloudflare Hosted Zone Record Type"
+  type        = string
+}
+
+variable "data" {
```
Comment (snovikov) [inline modules/record_collection_data/variables.tf:16]: It should be list of objects if you have multiple data records with the same name

File: variables.tf [hunk 5]
```diff
@@ -12,13 +12,17 @@ variable "records" {
   description = "List of names to create"
   type = list(object({
     name     = string
-    value    = string
+    value    = optional(string)
     type     = string
     ttl      = number
-    proxied  = bool
-    priority = number
+    proxied  = optional(bool)
```
Comment (snovikov) [inline variables.tf:18]: You have to address it also where you use this field. In case `proxied` is absent.

File: variables.tf [hunk 6]
```diff
@@ -12,13 +12,17 @@ variable "records" {
...
+    priority = optional(number)
```
Comment (snovikov) [inline variables.tf:19]: You have to address it also where you use this field. In case `priority` is absent.

**Review by snovikov — DISMISSED** [review]
There are some edge cases which are not covered with this PR. See my comments.

File: modules/record_collection_data/outputs.tf (hunk 1 above)
Comment (danvaida) [reply modules/record_collection_data/outputs.tf:4]: We can, of course.
And we should.
Every key should be something that can be looked up, but if you check the current implementation, it does the same thing with SPF records.
This way it is consistent.

File: main.tf (hunk 2 above)
Comment (danvaida) [reply main.tf:34]: I already tried this and only one data block is allowed.

File: modules/record_collection_data/main.tf (hunk 3 above)
Comment (danvaida) [reply modules/record_collection_data/main.tf:1]: Why is that? The module's iterative invocation happens in the main file.

File: modules/record_collection_data/variables.tf (hunk 4 above)
Comment (danvaida) [reply modules/record_collection_data/variables.tf:16]: > It should be list of objects if you have multiple data records with the same name

This will work as-is in such a case.

File: variables.tf (hunk 5 above)
Comment (danvaida) [reply variables.tf:18]: Not sure I understand what you mean...

File: variables.tf (hunk 6 above)
Comment (danvaida) [reply variables.tf:19]: > You have to address it also where you use this field. In case `priority` is absent.

Can you please elaborate or provide a failing test case?

**Comment by danvaida** [comment]
> There are some edge cases which are not covered with this PR. See my comments.

Please check my replies.

File: main.tf (hunk 2 above)
Comment (snovikov) [reply main.tf:34]: No, I mean not multiple blocks in one resource, but multimple records of type `CAA` with the same host name.
For example:

```
    {
      name     = "somedomain"
      type     = "CAA"
      ttl      = 120
      data     = {
        flags = 0
        tag   = "issue"
        value = "someca.com"
      }
    },
    {
      name     = "somedomain"
      type     = "CAA"
      ttl      = 120
      data     = {
        flags = 1
        tag   = "issue"
        value = "anotherca.com"
      }
    },
```

File: modules/record_collection_data/outputs.tf (hunk 1 above)
Comment (snovikov) [reply modules/record_collection_data/outputs.tf:4]: I see... that's true. This is a tricky one. In case of SPF/TXT records it has only value and if we replace `" "` with `_` we could have collisions. Let's say - you have two TXT records which have values:

- `word1 word2`
- `word1_word2`

if you replace spaces with undersocres you will have a collision. On the other hand - with `data` blocks you all the time in control of content and the key.

I don't see it as an huge issue. For consistency reasons - ignore my initial comment.

File: modules/record_collection_data/main.tf (hunk 3 above)
Comment (snovikov) [reply modules/record_collection_data/main.tf:1]: See my comments to `main.tf`

File: variables.tf (hunk 5 above)
Comment (snovikov) [reply variables.tf:18]: Should you use `lookup` function here, for example?

https://github.com/Flaconi/terraform-cloudflare-records/blob/a778d4ce612984678377dd090100ab405ded4c6c/main.tf#L17

File: variables.tf (hunk 6 above)
Comment (snovikov) [reply variables.tf:19]: Should you use `lookup` function here?

https://github.com/Flaconi/terraform-cloudflare-records/blob/a778d4ce612984678377dd090100ab405ded4c6c/main.tf#L18

**Review by snovikov — CHANGES_REQUESTED** [review]
I updated my comments

File: main.tf (hunk 2 above)
Comment (danvaida) [reply main.tf]: I see. Good catch. Fixed!

File: main.tf [hunk 7]
```diff
@@ -10,13 +12,35 @@ module "records" {
...
+}
+
+module "records_data" {
+  source   = "./modules/record_collection_data"
+  for_each = local.collections_data
+
+  zone_id = data.cloudflare_zone.this.id
+  name    = each.value[0].name
+  type    = each.value[0].type
+  ttl     = each.value[0].ttl
```
Comment (snovikov) [inline main.tf:34]: this will skip `ttl` and `proxied` for all records in `each.value`

File: variables.tf (hunk 5 above)
Comment (snovikov) [reply variables.tf:18]: solved

File: variables.tf (hunk 6 above)
Comment (snovikov) [reply variables.tf:19]: solved

**Review by snovikov — CHANGES_REQUESTED** [review]

**Review by snovikov — CHANGES_REQUESTED** [review]

File: modules/record_collection/versions.tf [hunk 8]
```diff
@@ -1,4 +1,5 @@
 terraform {
+  experiments = [module_variable_optional_attrs]
```
Comment (cytopia) [inline modules/record_collection/versions.tf:2]: Please also explicitly note the use of experimental TF features in the Readme.

**Review by snovikov — DISMISSED** [review]

**Review by cytopia — APPROVED** [review]

**Review by snovikov — APPROVED** [review]
